ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -1368,6 +1491,82 @@ private Offset getLastRecordOffset(LogSegment segment) {
     return lastOffset;
   }
 
+  // hardDeletePauseAndResume() helpers
+
+  /**
+   * Utility to test hard delete pause and resume with or without reload between pause and resume
+   * @param reloadIndex {@code true} if index has to be reloaded inbetween pause and resume. {@code false} otherwise
+   * @throws InterruptedException
+   * @throws IOException
+   * @throws StoreException
+   */
+  private void testHardDeletePauseResume(boolean reloadIndex) throws InterruptedException, IOException, StoreException {
+    properties.put("store.deleted.message.retention.days", Integer.toString(1));
+    properties.put("store.hard.delete.bytes.per.sec", Integer.toString(Integer.MAX_VALUE / 10));
+    properties.setProperty("store.data.flush.interval.seconds", "3600");
+    // enable daemon thread to run hard deletes
+    properties.put("store.enable.hard.delete", "true");
+    reloadIndex(false);
+    assertTrue("Hard delete is not enabled", index.hardDeleter.isRunning());
+    // IndexSegment still uses real time so advance time so that it goes 2 days past the real time.
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    long expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // give it some time so that hard delete completes one cycle
+    waitUntilExpectedProgress(expectedProgress, 5000);
+    verifyEntriesForHardDeletes(deletedKeys);
+
+    Set<MockId> idsDeleted = new HashSet<>();
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    Set<MockId> idsToDelete = new HashSet<>();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    idsDeleted.addAll(idsToDelete);
+    // pause hard delete
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    waitUntilExpectedState(Thread.State.WAITING, HardDeleter.HARD_DELETE_SLEEP_TIME_ON_CAUGHT_UP + 1, 10);
+
+    // delete two entries
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.clear();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    idsDeleted.addAll(idsToDelete);
+
+    if (reloadIndex) {
+      reloadIndex(true);
+      assertEquals("Hard deletes should have been paused ", Thread.State.WAITING, index.hardDeleteThread.getState());
```

Comment:
I think you should still do `waitUntil` here because the thread does start on reloading the index and stops only when it gets to the pause code (which should be quick but is not instant)